content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -5,6 +5,84 @@ prev: /recipes/authoring/kata-snippets/sample-tests/
+- **Provide both [fixed](#fixed-tests) and [random](#random-tests) tests** for the submit tests. Some test frameworks offer features like test case generators (`QuickCheck`, `ScalaCheck`), and when used, authors can choose different structures for their test suite, but still they should create test specifications covering all possible scenarios. See specific sections below for more details.
+- **Tests should not be susceptible to input mutation.** It should not be possible for the user solution to modify its input in a way which would influence the execution, result, or output of a test. If a reference solution is used, it should be called first, and the user solution should be called afterwards with the same input. Potentially mutated input should not be used to compose any information presented to the user.
+- **Assertions should be placed near invocations of the user's solution**, preferably immediately following it. Debugging becomes difficult for the user when a significant amount of post-processing occurs between the result returned by the user's solution and the corresponding assertion(s).
+- **The test suite should not take a long time to run.** For kata which are not performance-oriented, the test suite should complete well below the time limit, ideally within 3 seconds. There's a time limit of 12 seconds (16 or 20 seconds for a few languages) for every attempt, but the generation and execution of test cases including any calls to the user (and reference) solution should run well below this limit. Otherwise, there's a risk that tests will inconsistently time out, or some valid, but not optimally performant solutions will be rejected.
+
+
+## Reference Solution
+
+Some test suites require a reference solution to generate the expected value(s) and compare them to those returned by the user solution. While such practice is common and sometimes necessary, there are things which have to be handled carefully:
+
+- **Avoid using a reference solution if at all possible.** Sometimes there's no other way, but quite often a reference solution is simply not necessary. Most of the time, test cases can be generated with the answer known upfront which eliminates an entire class of potential problems associated with reference solutions:
+  - Extra time spent on running a reference solution
+  - Reference solution being accessible to users by mistake
+  - Input mutation by the user solution affecting the input passed to the reference solution
+  - Incorrect implementation of the reference solution leading to rejection of valid user solutions
+- **Reference solution should not be revealed to the user.** When an assertion fails or the test suite crashes, some testing frameworks print fragments of source code which caused the failure to the console. It may happen that such printed failure messages or stack traces expose information about the solution which should not be revealed.
```

Comment:
```suggestion
- **The reference solution should not be revealed to the user.** When an assertion fails or the test suite crashes, some testing frameworks print fragments of source code which caused the failure to the console. It may happen that such printed failure messages or stack traces expose information about the solution which should not be revealed.
```